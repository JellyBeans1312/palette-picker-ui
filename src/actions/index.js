export const addProject = id => ({
  type: 'ADD_PROJECT',
  id
});

export const addAllProjects = allProjects => ({
  type: 'ADD_ALL_PROJECTS',
  allProjects
});

export const saveColors = colors => ({
  type: 'SAVE_COLORS',
  colors
});