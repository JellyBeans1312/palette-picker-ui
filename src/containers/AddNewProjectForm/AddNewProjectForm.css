.new-project-form {
  background-color: blue;
  height: 50%;
  width: 100%;
}