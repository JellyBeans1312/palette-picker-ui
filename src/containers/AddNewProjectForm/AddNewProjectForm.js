import React, { Component } from 'react';
import './AddNewProjectForm.css'
import { createProject, getAllProjects } from '../../util/apiCalls';
import { addProject, addAllProjects } from '../../actions';
import { connect } from 'react-redux';

export class AddNewProjectForm extends Component {
  constructor() {
    super();
    this.state = {
      projectName: '',
      error: ''
    }
  }

  handleChange = e => {
    this.setState({[e.target.name]: e.target.value})
  }

  handleSubmit = async e => {
    e.preventDefault();
    const newProject = {
      project_name: this.state.projectName
    }
    try {
      const project = await createProject(newProject);
      this.props.addProject(project);
    } catch (error) {
      this.setState({ error: error.message })
    }
    const allProjects =  await getAllProjects();
    this.props.addAllProjects(allProjects);
    this.setState({projectName: ''});
  }

  render() {
    return (
      <form> 
        {this.state.error && <p>{this.state.error}</p>}
        <input 
          type='text'
          name='projectName'
          placeholder='Enter a new project name!'
          onChange={this.handleChange}
          value={this.state.projectName}
        />
        <button onClick={this.handleSubmit}>Create Project</button>
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addProject: project => dispatch(addProject(project)),
  addAllProjects: allProjects => dispatch(addAllProjects(allProjects))
})

export default connect(null, mapDispatchToProps)(AddNewProjectForm);