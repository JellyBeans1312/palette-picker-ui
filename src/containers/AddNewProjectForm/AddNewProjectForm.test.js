import React from 'react';
import { shallow } from 'enzyme';
import { addProject, addAllProjects } from '../../actions/index';
import { AddNewProjectForm, mapDispatchToProps } from './AddNewProjectForm';

describe('AddNewProjectForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<AddNewProjectForm />)
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state when a change is detected in the input', () => {
    let mockEvent = {
      target: {
        name: 'projectName',
        value: 'New Project'
      }
    }

    wrapper.find('input').simulate('change', mockEvent);

    expect(wrapper.state('projectName')).toEqual('New Project');
  });

  it('should call handleSubmit when the Create Project button is clicked', () => {
    let mockEvent = {
      preventDefault: jest.fn()
    }

    wrapper.instance().handleSubmit = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('button').simulate('click', mockEvent);

    expect(wrapper.instance().handleSubmit).toHaveBeenCalledWith(mockEvent);
  });

  describe('mapDispatchToProps', () => {
    it('should dispatch with a project_name and id when addProject is called', () => {
      let mockProjectName = 'New Project';
      let mockId = 1;
      let mockDispatch = jest.fn();

      let actionToDispatch = addProject(mockProjectName, mockId);
      let mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addProject(mockProjectName, mockId);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should dispatch with allProjects when addAllProjects is called', () => {
      let mockAllProjects = [
        {
          id: 51,
          project_name: "Aidan's fantastic project"
        },
        {
          id: 52,
          project_name: "Eric's great project"
        }
      ]
      let mockDispatch = jest.fn();

      let actionToDispatch = addAllProjects(mockAllProjects);
      let mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addAllProjects(mockAllProjects);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});