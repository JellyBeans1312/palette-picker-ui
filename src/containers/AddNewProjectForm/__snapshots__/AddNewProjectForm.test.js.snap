// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewProjectForm should match the snapshot with all data passed in correctly 1`] = `
<form
  className="new-project-form"
>
  <div
    className="align-form-items"
  >
    <input
      className="project-name"
      name="projectName"
      onChange={[Function]}
      placeholder="Enter a new project name!"
      type="text"
      value=""
    />
    <button
      className="create-project-button"
      onClick={[Function]}
    >
      Create Project
    </button>
  </div>
</form>
`;
